### ISSUE TITLE
Ubuntu20 checksum in packer json is incorrect

### ISSUE THREAD

**Issue by walterrowe** [geerlingguy/packer-boxes#52]

**Comment by mkinney** [comment]
I'm guessing because they do daily builds of the iso until it is released on April 23, 2020.

**Comment by walterrowe** [comment]
Makes sense. I am new to packer, so it forced me to learn a little. I updated the checksum and it ran fine. Feel free to close this, or keep it until the official release is out.

**Comment by geerlingguy** [comment]
@walterrowe - Yeah, sorry about that. Unfortunately short of me writing a script to update it nightly, you'll have to live with it until the official 20.04.0 release comes out :(

I'll leave this open for visibility, though, until the release comes (should be soon!)

**Comment by geerlingguy** [comment]
Ubuntu 20.04 is here: https://releases.ubuntu.com/20.04/

So I'll update to the latest stable ISO and then we can close out this issue :)

**Comment by geerlingguy** [comment]
So... I can only find a server live install ISO, and the 'server-amd64' isos are stored under 'daily' and are listed as "legacy" for some reason: http://cdimage.ubuntu.com/ubuntu-server/daily/20200423/

I'm trying a build with one of those right now. The 'live-server' install ISO doesn't behave the same and I can't get it to go into kickstart mode for installation... don't know why it's so different than the 'legacy' ISO.

**Comment by geerlingguy** [comment]
Oh. https://discourse.ubuntu.com/t/server-installer-plans-for-20-04-lts/13631

> With 20.04 LTS, we will be completing the transition to the live server installer and discontinuing the classic server installer based on debian-installer (d-i), allowing us to focus our engineering efforts on a single codebase. The next-generation subiquity server installer brings the comfortable live session and speedy install of Ubuntu Desktop to server users.

**Comment by geerlingguy** [comment]
So, the new auto-installer is documented here: https://wiki.ubuntu.com/FoundationsTeam/AutomatedServerInstalls

And a Quick Start guide is available here: https://wiki.ubuntu.com/FoundationsTeam/AutomatedServerInstalls/QuickStart

Though it seems like it's a little different if booting the ISO in Packer—it looks like the quick start guide assumes you're building a new image / install from within an existing instance, not building on top of something that's already running.

In other words, I give up for now on moving to this new installer—I'll likely take a look again soon, but for now, I'll leave up the 0.9.0 box with a few weeks-old nightly build, because it's going to be at least a couple hours' work to try to figure out this new unattended installation system.

**Comment by geerlingguy** [comment]
The official Packer documentation probably needs to be updated too: https://www.packer.io/guides/automatic-operating-system-installs/preseed_ubuntu.html

Bento hasn't been updated so I can't steal any ideas from there yet, either: https://github.com/chef/bento/tree/master/packer_templates/ubuntu

**Comment by geerlingguy** [comment]
For now, I'm sticking with a legacy daily server image build: http://cdimage.ubuntu.com/ubuntu-server/daily/20200423/focal-legacy-server-amd64.iso

**Comment by walterrowe** [comment]
Bring up your Ubuntu 20.04 vagrant instance and look in the install logs dir there. You will find an auto-install file that matches the anaconda configs for that machine. If I read the docs correctly, you can use that auto-install file for the auto-install method. You still need to figure out how to initiate an auto-install vs d-i, and point it to the auto-install file. It is basically YAML format.

**Comment by geerlingguy** [comment]
@walterrowe - you will see install screens in vBox as they play

Yeah, was doing that and I was trying to manually get into an automated install mode but was unsuccessful after a few minutes digging around. But I'll likely get around to it again soon.

**Comment by tmepple** [comment]
@geerlingguy Instead of using the daily link pinned at 20200423 (which they might prune) should we use the build at http://cdimage.ubuntu.com/ubuntu-legacy-server/releases/20.04/release/ubuntu-20.04-legacy-server-amd64.iso?

I assume they plan to keep releasing legacy versions using d-i through the 20.04 lifetime so we have plenty of time to transition to the new installer.

**Comment by walterrowe** [comment]
According to the [Installer section of the 20.04 LTS release notes](https://wiki.ubuntu.com/FocalFossa/ReleaseNotes?_ga=2.101841992.873863106.1587665863-1943852030.1586865872#Installer), the Live Installer is now the preferred install method.

**Comment by basictheprogram** [comment]
Just fyi, http://cdimage.ubuntu.com/ubuntu-server/daily/20200423/focal-legacy-server-amd64.iso inside the box-config.json gives me a 404 error but http://cdimage.ubuntu.com/ubuntu-legacy-server/releases/20.04/release/ubuntu-20.04-legacy-server-amd64.iso seems to work for me.

**Comment by nickcharlton** [comment]
I've been looking into this problem, and after a weekend of experiments have gotten it working using the new automated installer.

It's a bit late here in the UK as I write this, so I've run out of time to open the PR I wanted to but I'd like to come back tomorrow with one.

In the mean time, if someone wants to take a look at this, I've written up my notes which have a working configuration: https://nickcharlton.net/posts/automating-ubuntu-2004-installs-with-packer.html

**Comment by geerlingguy** [comment]
@nickcharlton - Awesome, thanks for posting! For the shorter-term I merged @basictheprogram's PR which uses the more stable (non-nightly) legacy build ISO.

**Comment by rubenst2013** [comment]
@nickcharlton how did you get the installer to execute the user-data yaml formated file?
I did as you described in your blog post but its simply stuck at language selection.
I noticed once it reaches the first interactive screen with language selection there is no hot key f6...only f2 to jump into the command line. 

Increased the boot wait time, because at 5s it was typing #somewhere(tm) in the middle of the installer startup.
And of course added the vars for the http server and port. No luck.

**Comment by nickcharlton** [comment]
@rubenst2013, I was seeing random failures with that problem, too. Sometimes Packer would try and connect, get a failure and quit. Other times it'd finish fine.

If you try using [`pause_before_connecting`](https://www.packer.io/docs/builders/virtualbox-iso.html#pause_before_connecting), does that help?

**Comment by rubenst2013** [comment]
Hi @nickcharlton, 

Thanks for the heads-up about this setting. Though I just tried it and the VM was closed by packer again, regardless of that setting. :cry:

**Comment by rubenst2013** [comment]
I made a pull request, which implements the necessary changes and worked locally. 

Please let me knwo how this works for you. Thanks! 

# #61

**Comment by nickcharlton** [comment]
Right, yeah. I found the same yesterday too. `pause_before_connecting` doesn't behave the way I thought it did: It doesn't wait for SSH, instead it keeps trying to connect and when it eventually can it _then_ pauses.

Instead, I found that [expanding the `ssh_handshake_attempts` achieves the right thing](https://github.com/nickcharlton/packer-ubuntu-2004/blob/051588296fc53bf742dd59fb1a8599faeb2ee98e/ubuntu-2004.json#L16), by allowing it to keep failing longer.

**Comment by cypherphunk** [comment]
@geerlingguy As a little help, I managed to get this going via cloud-init. Ubuntu 20.04 pickups the user-data correctly. Take a look at my [repo](https://gitlab.com/infrastructure-as-code1/packer/-/blob/master/ubuntu-20.04-lts.json), especially at boot_command and ssh_handshake_attempt (I set it to 40, but maybe it could be set lover - didn't had time to test it)

**Comment by walterrowe** [comment]
Is this fixed / closed?

**Comment by geerlingguy** [comment]
@walterrowe - I've been using my workaround since the beginning—but I'd still like to go back to the smaller ISO and maybe save some space and setup time, so I'm leaving it open :)

**Comment by echeadle** [comment]
Hi guys.  I have been working on using packer with  the new hcl files formats and with subiquity.  It has been quite a challenge.  I don't know who came up with the boot_command, but I have seen no one agree on how it should look.  There are bunches of ways of doing it wrong.

I used the packer hcl2_upgrade command and a whole of of googling to figure out what to do.

I have been researching the issue for several days.  Nick's information was the invaluable, but I was still having problems until I changed the way I encrypted the password and ran across someone who stopped the ssh server using the early-command section.

Anyway I have converted the json file to HCL and the preseed file to user-data for cloud-init.  It now seems to be reliably working either way.  

The code is in the repo: 
                         https://github.com/echeadle/packer-tpls.git

Essentially I started with Jeff's packer-boxes repo, removed everything but Ubuntu 20.04 and started to work through the files.  I first ran the json version after countless tries at revamping the user-data files.  I tried many different things, but finally ended up with Nick's user-data file and added:

  early-commands:
    - sudo systemctl stop ssh

I also added
 late-commands:
  - sed -i -e 's/^#\?PasswordAuthentication.*/PasswordAuthentication yes/g' /target/etc/ssh/sshd_config
    - sed -i -e 's/^#\?PermitRootLogin.*/PermitRootLogin yes/g' /target/etc/ssh/sshd_config
    - echo 'vagrant ALL=(ALL) NOPASSWD:ALL' > /target/etc/sudoers.d/vagrant
    - sudo systemctl start ssh

 I probably did not need to add the Password Authentication or the PermitRootLogin, but I did need to restart ssh.  I put those other commands in, because I was having so many issues with logging into the box.  I will be testing more and I suspect you added the vagrant sudo line to sudoers, but I was just trying to get it to work.

I have run packer by hand and with ansible and it seems to be working fine.

I have Jeff's Ansible and Kubernetes books and find them fascinating. I saw tJeff's repo and wanted to get it working, but packer's change to hcl 2 and Ubuntu getting rid of the preseed method was a problem, so I thought I would try and get it working. With the info I found this issue discussion,  it was the final piece that got everything working, so I thought I would post twhat I had changed in a  repo for easy access. 

If you find issues, please let me know.  I will be refining the code over the next few days.

**Comment by echeadle** [comment]
I was too quick to claim success.  The code is working, it pushes a new box to the vagrant website, but when I run vagrant up, it looks like cloud-init is still doing something and nfs does not work.  I have more work to do.

**Comment by echeadle** [comment]
The code seems to be working. The problem with removing cloud-init with ,  commands, is that it removes cloud-init before installation.  Shows my inexperience with cloud-init.

Also I made a rookie error when moving the clean up of cloud-init to the cleanup.sh script;  I did not stop the service. 

So I added stopping and disabling the systemd service cloud-init, then apt-get purge works.

I am having issues with nfs, which is why I am running my vagrant file without it.  But I think that is an issue with my laptop, but ive not been able to test that feature.

I hope this helps.

**Comment by echeadle** [comment]
Everything is working now for ubuntu 20.04. and I was able to get NFS to work,

So now I am going to try 21.10.  I started looking at the desktop system, but ubiquity looks completely different and I don't feel like tackling it right now.  I prefer to try and duplicate the work on the next version of the server OS.

**Comment by echeadle** [comment]
The boot_command found in the Bento repository works with 21.10.  With this boot_command and the packer hcl_upgrade command it is easy to use HCL templates now.

**Comment by maroskukan** [comment]
Hey folks. would you still be interested in "live installer" template for Ubuntu 20.04? I have something like this on the radar. Perhaps I could do a PR afterwards.
